packages/yew/src/dom_bundle/subtree_root.rs

```diff
@@ -0,0 +1,513 @@
+//! Per-subtree state of apps
+
+use super::{test_log, Registry};
+use crate::virtual_dom::{Listener, ListenerKind};
+use gloo::events::{EventListener, EventListenerOptions, EventListenerPhase};
+use std::cell::RefCell;
+use std::collections::HashSet;
+use std::hash::{Hash, Hasher};
+use std::rc::{Rc, Weak};
+use std::sync::atomic::{AtomicBool, AtomicU32, Ordering};
+use wasm_bindgen::prelude::wasm_bindgen;
+use wasm_bindgen::JsCast;
+use web_sys::{Element, Event, EventTarget as HtmlEventTarget};
+
+/// DOM-Types that capture (bubbling) events. This generally includes event targets,
+/// but also subtree roots.
+pub trait EventGrating {
+    fn subtree_id(&self) -> Option<TreeId>;
+    fn set_subtree_id(&self, tree_id: TreeId);
+    // When caching, we key on the length of the `composed_path`. Important to check
+    // considering event retargeting!
+    fn cache_key(&self) -> Option<u32>;
+    fn set_cache_key(&self, key: u32);
+}
+
+#[wasm_bindgen]
+extern "C" {
+    // Duck-typing, not a real class on js-side. On rust-side, use impls of EventGrating below
+    type EventTargetable;
+    #[wasm_bindgen(method, getter = __yew_subtree_id, structural)]
+    fn subtree_id(this: &EventTargetable) -> Option<TreeId>;
+    #[wasm_bindgen(method, setter = __yew_subtree_id, structural)]
+    fn set_subtree_id(this: &EventTargetable, id: TreeId);
+    #[wasm_bindgen(method, getter = __yew_subtree_cache_key, structural)]
+    fn cache_key(this: &EventTargetable) -> Option<u32>;
+    #[wasm_bindgen(method, setter = __yew_subtree_cache_key, structural)]
+    fn set_cache_key(this: &EventTargetable, key: u32);
```

Comment:
I assume we use these properties for caching purposes? Why not keep them on the Rust side?

Reply:
Those have to survive js-land. They are accessed during event handling before it's clear which tree is responsible. There are two objects: on `HtmlNode`s, the `subtree_id` is responsible for branding the element, so we can find the responsible subtree, `cache_key` is unused. On `Event`s, it's purely for caching. Both need to be accessed early during event handling, where no state in rust is established, yet, basically at the start of the callgraph.